graphistry/collapse.py

```diff
@@ -0,0 +1,532 @@
+    ex:
+        node = "4" will take any sequence from get_cluster_store_keys, "1 2 3", "4 3" and returns "1 2 3 4"
+        when they have a common entry (3).
+
+    -------------------------------------------------------------------------------------------------------------
+
+    :param ndf, node DataFrame
+    :param node: node to melt
+    :returns new_parent_name of super node
+    """
+    """MAIN AWESOME"""
+    # suppose node = "4" will take any sequence from get_cluster_store_keys, "1 2 3", "4 3" and returns "1 2 3 4"
+    rdf = ndf[get_cluster_store_keys(ndf, node)]
+    topkey = wrap_key(node)
+    if not rdf.empty:
+        for key in rdf[COLLAPSE_NODE].values:  # all these are already wrapped
+            # print(f'collapse key iter: {key}')
+            # add the keys to cluster
+            topkey += f" {key}"  # keep whitespace
+    topkey = reduce_key(topkey)
+    if VERBOSE:
+        print(f"start of collapse: {node}")
+        print(f"Collapse Final Key: {topkey}")
+    return topkey
+
+
+def get_new_node_name(ndf, parent, child) -> str:
+    """
+        If child in cluster group, melts name, else makes new parent_name from parent, child
+
+    ---------------------------------------------------------------------------------------------------------
+
+    :param ndf: node DataFrame
+    :param parent: `node` with `attribute` in `column`
+    :param child: `node` with `attribute` in `column`
+    :returns new_parent_name
+    """
+    # THIS IS IMPORTANT FUNCTION -- it is where we wrap the parent/child in WRAP
+    # if child in cluster group, we melt it
+    ckey = in_cluster_store_keys(ndf, child)
+
+    if ckey:
+        new_parent_name = melt(ndf, child)
+    else:  # if not, then append child to parent as the start of a new cluster group
+        # might have to escape parent and child if node names are dumb eg, 'this value key'
+        new_parent_name = f"{wrap_key(parent)} {wrap_key(child)}"
+    if VERBOSE:
+        print(f"Renaming parent {parent} with child {child} as {new_parent_name}")
+    return reduce_key(new_parent_name)
+
+
+def collapse_nodes(g, parent, child):
+    """
+        Asserts that parent and child node in ndf should be collapsed into super node.
+        Sets new ndf with COLLAPSE nodes in graphistry instance g
+
+        # this asserts that we SHOULD merge parent and child as super node
+        # outside logic controls when that is the case
+        # for example, it assumes parent is already in cluster keys of COLLAPSE node
+
+    ---------------------------------------------------------------------------------------
+
+    :param g: graphistry instance
+    :param parent: `node` with `attribute` in `column`
+    :param child: `node` with `attribute` in `column`
+    :returns: graphistry instance, new_parent_name of super cluster
+    """
+
+    ndf, edf, src, dst, node = unpack(g)
+
+    new_parent_name = get_new_node_name(ndf, parent, child)
+
```

Comment:
hmmm this is a mutation... probably better not to return anything

Reply:
Fixed